### DISCUSSION TITLE
How to implement non-stringable runtime configuration for server handlers provided by module?

### DISCUSSION THREAD

**Discussion by dulnan** [nuxt/nuxt#15898] [Questions]
Let's say I implement a module that adds a server handler. The server handler makes a request to an external API.
I would like to offer users of my module to provide a `beforeRequest` function, which I can call in the server handler.

Currently this callback can be defined in my module's options, so I have some boilerplate code to load the module config during runtime. For that I need to know the root directory when calling `loadNuxtConfig()`, which I determine in my module's setup and provide via runtime config. I only realized later that this of course doesn't work the moment the root directory is different during build and runtime.

Generally unhappy with my approach, I researched a bit for better solutions. For plugins the suggestion (#9192) is to use app.config.ts, but this doesn't work in server handlers. Another way seems to be to create a custom configuration file, like is done with `router.options.ts` in the pages module, where users can create such a file in their app to provide the runtime function.

What other possible approaches are there and which ones are considered "best practice"?

**Comment by Baroshem** [comment]
I encountered similar problem with my security module for nuxt. I wanted to allow users to pass a custom method to handle case where certain security middleware is triggered so that they can add a custom functionality after certain condition is met (for example user used wrong HTTP method). 

Unfortunately, I have not found any useful approach here, so decided to recommend that in such case, users should just create a similar endpoint from scratch. This is definitely not a good practise, as I would love to be able to give users an option to just pass a custom method but have not found a way to do this efficiently.

I will hook to this thread so maybe someone will find an answer for that

**Reply by dulnan** [reply]
I spent two days looking through the source of dozens of Nuxt 3 modules and Nuxt itself and I think I have finally found a reliable approach to offer module users a way to provide config containing functions. Accessing the config in a server handler works, without having to load nuxt.config.ts!

My first tests were successful, I then implemented it in one of my modules and all my tests pass, so I think it works!

Here's a working example with the minimum code required to implement it:
https://stackblitz.com/edit/nuxt-starter-qzqwkx?file=nuxt.config.ts,app.vue,app%2Ffoobar.options.ts&file=modules%2Ffoobar%2Fruntime%2FserverHandler%2Ftest.ts

Basically I define a location for a config file at `~/app/foobar.options.ts` where users can put the special config in their Nuxt app. The module then does `addTemplate()`, with the following contents:

```typescript
export { default } from '/app/foobar.options.ts'
```

The key is to then inline the template:

```typescript
nuxt.options.nitro.externals.inline.push(template.dst)
```

And define an alias:
```typescript
nuxt.options.alias['#foobar-server-options'] = template.dst
```

Now we can import the file in our server handler:

```typescript
import { defineEventHandler } from 'h3';
import runtimeModuleOptions from '#foobar-server-options';

export default defineEventHandler(async (event) => {
  return await runtimeModuleOptions.getData();
});
```

In the generated `node-server.mjs` the contents of `foobar.options.ts` are inlined.

I also created a typed helper function `defineFoobarServerOptions` which is exported from the module.

So, I still don't know if this is considered a "good" approach, but I'm just glad that it works and it seems much more reliable/sane than calling `loadNuxtConfig()` at runtime. I'll do some more testing next week with some edge cases and then see how it goes!
